src/Covenant/Type.hs

```diff
@@ -706,3 +713,132 @@ substitute index toSub = \case
+boundIdents :: forall (ann :: Type). Lens' (PrettyContext ann) (Map ScopeBoundary [Doc ann])
+boundIdents = lens goGet goSet
+  where
+    goGet (PrettyContext bi _ _) = bi
+    goSet (PrettyContext _ scop vars) bi' = PrettyContext bi' scop vars
+
+currentScope :: forall (ann :: Type). Lens' (PrettyContext ann) ScopeBoundary
+currentScope = lens goGet goSet
+  where
+    goGet (PrettyContext _ scop _) = scop
+    goSet (PrettyContext bi _ vars) scop  = PrettyContext bi scop vars
+
+varStream :: forall (ann :: Type). Lens' (PrettyContext ann) [Doc ann]
+varStream = lens goGet goSet
+  where
+    goGet (PrettyContext _ _ vars) = vars
+    goSet (PrettyContext bi scop _) = PrettyContext bi scop
+
+-- Generate N fresh var names and use the supplied monadic function to do something with them.
+withFreshVarNames :: forall (ann :: Type) (a :: Type). Int -> ([Doc ann] -> PrettyM ann a) -> PrettyM ann a
+withFreshVarNames n act = do
+  stream <- asks (view varStream)
+  let (used,rest) = splitAt n stream
+  local (set varStream rest) $ act used
+
+-- Maybe make a newtype with error reporting since this can fail, but do later since *should't* fail
+newtype PrettyM (ann :: Type) (a :: Type) = PrettyM (Reader (PrettyContext ann) a)
+  deriving newtype (Functor,Applicative,Monad,MonadReader (PrettyContext ann))
+
+runPrettyM :: forall (ann :: Type) (a :: Type). PrettyM ann a -> a
+runPrettyM (PrettyM ma) = runReader ma (PrettyContext mempty 0 infiniteVars)
+
+
+-- REVIEW @Koz: You said we always consider a boundary crossed even if the count is zero,
+--              but when I do that it blows up lists? I think I have to be misunderstanding something
+bindVars :: forall (ann :: Type) (a :: Type). Count "tyvar" -> ([Doc ann] -> PrettyM ann a) -> PrettyM ann a
+bindVars count' act
+  | count == 0 = crossBoundary (act [])
+  | otherwise = crossBoundary $ do
+      here <- asks (view currentScope)
+      withFreshVarNames count $ \newBoundVars ->
+        local (over boundIdents (Map.insert here newBoundVars)) (act newBoundVars)
+ where
+   -- Increment the current scope
+   crossBoundary :: PrettyM ann a -> PrettyM ann a
+   crossBoundary = local (over currentScope (+ 1))
+
+   count :: Int
+   count = review intCount count'
+
+-- Bad name, but anyway, looks up the Doc for a pretty
+lookupRigid :: forall (ann :: Type). Int -> Index "tyvar" -> PrettyM ann (Doc ann)
+lookupRigid (ScopeBoundary -> scopeOffset) argIndex = do
+  let argIndex' = review intIndex argIndex
+  here <- asks (view currentScope)
+  asks (preview (boundIdents % ix (here + scopeOffset) % ix argIndex')) >>= \case
+    Nothing -> -- TODO: actual error reporting
+               error $ "Internal error: The encountered a variable at arg index "
+                       <> show argIndex'
+                       <> " with true level "
+                       <> show scopeOffset
+                       <> " but could not locate the corresponding pretty form at scope level " <> show here
```

Comment:
Can this ever actually fail in practice?

Reply:
As things are now? Probably not. The problem arose with empty scope boundaries in the builtin nested List type. I'd prefer to leave it here though since eventually we will probably have examples which can actually make it fail.